src/storage/src/source/reclock/compat.rs

```diff
@@ -135,13 +124,13 @@ where
         use futures::stream;
         let events = stream::once(async move {
             let updates = read_handle
-                .snapshot_and_fetch(as_of.clone())
```

Comment:
Why are we not using the `as_of` here since we are advancing the timestamps to that frontier below anyway?

Reply:
Same concern as here, basically: https://github.com/MaterializeInc/materialize/pull/27314#discussion_r1620860314

Another possibly-more-readable option would be to start reading from the greater of the two frontiers.

Reply:
(Pushed a commit with that idea; LMK what you think.)

Reply:
Ah, I see. That makes sense but I don't think we should have this logic here. The `as_of` parameter basically communicates "the collection accumulates correctly as of this frontier" but that's not true in the cases you mentioned. This leads me to the conclusion that the `as_of` should never have had this value in the first place. My suggestion would be to change the async storage worker to set as_of to `(upper - 1).join(since)`

Reply:
> The as_of parameter basically communicates "the collection accumulates correctly as of this frontier" but that's not true in the cases you mentioned.

I think it might be true! Here's the reasoning:
- In normal operation, the new as-of should be past the since, and so if it accumulated correctly at the since it should accumulate correctly at the as-of.
- On start-up, the normal relationship is inverted... the as-of is at `[0]` and the since is at some greater frontier `F > [0]`. Normally, we'd have no reason to think that accumulating at `[0]` would give the same result as accumulating at `F`. But _in the case of the source specifically_, we happen to know that the collection will be empty between `[0` and `[F]`, since we don't reclock data before our initial since, so the fact that we start reading later doesn't matter.
    - It seems gross taking advantage of this special property, but note that the source is _already_ taking advantage of this special property... if this wasn't true, and we could end up reclocking input data to some time before `F` in our source exports, then accumulating up to `F` would give bad results.

Reply:
> On start-up, the normal relationship is inverted... the as-of is at `[0]` and the since is at some greater frontier `F > [0]`

This is what doesn't sound right to me. Each collection has an associated `since` frontier, which defines the minimum times beyond which you can accumulate to correct results. When a dataflow is rendered that computes a result based on some input we can define the set of times that we're interested in for the result. This is what an `as_of` frontier communicates and it only makes sense when it's greater than or equal to the since frontier. It basically says "I know you have all this data since the `since` frontier, but I only want to compute the states as of the `as_of` frontier, so please advance all the times and compact them so I do less work".

> But in the case of the source specifically, we happen to know that the collection will be empty between [0 and [F],

It is undefined to talk about the contents of a collection for times that are not beyond its since frontier so I'm not sure what it means that we happen to know its contents before the since frontier F.

That said, the remap collection couldn't possibly be empty between 0 and F, if that was the case then that would mean that at timestamp 0 the remap collection would accumulate to the empty frontier, which would mean that the entirety of the source data would need to be reclocked to that time. But again, this is a bit meaningless when the since frontier disallows talking about those times.

Is something broken if we change the async storage worker to say "we want to read the remap collection as of its since frontier or upper - 1, whichever is greater"?

Reply:
> That said, the remap collection couldn't possibly be empty between 0 and F, if that was the case then that would mean that at timestamp 0 the remap collection would accumulate to the empty frontier, which would mean that the entirety of the source data would need to be reclocked to that time.

Good callout, thank you!

> But again, this is a bit meaningless when the since frontier disallows talking about those times.

Agree that it's not particularly meaningful to talk about the contents of the _shard_, but I think it's fair to talk about the contents of the "logical collection", as you've just shown! 😅

Anyways - I'll see if I can do one more take on this, based only on "stable" information about the outputs, and report back.

Reply:
Okay, I think the latest version maybe nobody will hate!

The idea here is to use the since of the _export_ collection(s), not the remap collection. The general idea is that we if nobody will ever read any of our exports as-of some time, we don't need to reproduce our export as-of that time, and thus we don't need to read the remap collection as-of that time. This helps in the specific problem case on bootstrap, where the controller had advanced the since well past the upper... but it also doesn't rely on any particular ordering of downgrades in the controller, which solves my problem.

Reply:
> The general idea is that we if nobody will ever read any of our exports as-of some time, we don't need to reproduce our export as-of that time, and thus we don't need to read the remap collection as-of that time. 

That would indeed the most correct statement! However, where do we ensure that by the time this async worker reads the since frontier of an export it is not still at `{0}`? There have been many refactors in the controller and I couldn't trace where the since handle is used. Is the controller ensuring that shard initialization has finished before sending the command to the cluster?

Reply:
Two answers:
- [`StorageController::create_collections`](https://github.com/MaterializeInc/materialize/blob/52ee3f0fe46d23869274d6d7d8b47aaf48a7aab0/src/storage-client/src/storage_collections.rs#L1296) appears to downgrade the since handles as part of `open_data_handles`, which should occur before the actual exports are created. Though if @aljoscha can confirm this that is helpful!
- If the since was only advanced asynchronously, I believe we'd already have problems: since any reader that could read the shard before that initial since advancement could hold back the since, possibly reading data as of a time before the source was created and thus before we could possibly have a correct result. And if so this property is already required for correctness...

Reply:
I think what I may be remembering is the time before Sean fixed the subsource dependency inversion thing where the controller would first learn about subsources (with no context on which source they belong to) and only afterwards learn about the source itself.

> And if so this property is already required for correctness...

I think correctness property #2 is not working for `ADD SUBSOURCE` for a bunch of reasons that need.. (you guessed it) some of the controller refactors to land.

Cool, then the current state looks good to me!

Reply:
Thanks for working through it with me! A bit of a winding road but I'm pretty happy with where we ended up.

I'll get this in to let it bake a bit before the release cut this week...